File: l3/z3/DHSetup.cs
using System;
using System.Collections.Generic;

public class DHSetup<T> where T : IFieldElement<long>, new()
{
    private T generator;

    public DHSetup()
    {
        Random random = new Random();
        this.generator = GenerateValidGenerator(random) ?? throw new InvalidOperationException("Failed to generate a valid generator");
    }

    private T GenerateValidGenerator(Random random)
    {
        T element;
        List<long> primeDivisors = CalculatePrimeDivisors(new T().GetCharacteristic() - 1);

        while (true)
        {
            long candidateValue = 2 + random.Next((int)(new T().GetCharacteristic() - 2));
            element = (T)Activator.CreateInstance(typeof(T), candidateValue)!;

            bool isValidGenerator = true;
            foreach (long prime in primeDivisors)
            {
                T powerResult = Power(element, (new T().GetCharacteristic() - 1) / prime);
                if (powerResult.Value == 1)
                {
                    isValidGenerator = false;
                    break;
                }
            }

            if (isValidGenerator)
            {
                return element;
            }
        }
    }

    private List<long> CalculatePrimeDivisors(long n)
    {
        List<long> primeDivisors = new List<long>();
        while (n % 2 == 0)
        {
            primeDivisors.Add(2);
            n /= 2;
        }
        for (long i = 3; i <= Math.Sqrt(n); i += 2)
        {
            while (n % i == 0)
            {
                primeDivisors.Add(i);
                n /= i;
            }
        }
        if (n > 2)
        {
            primeDivisors.Add(n);
        }
        return primeDivisors;
    }

    public T GetGenerator()
    {
        return generator;
    }

    public T Power(T a, long b)
    {
        T result = new T();
        T baseValue = a;
        while (b > 0)
        {
            if (b % 2 == 1) result = (T)result.Multiply(baseValue);
            baseValue = (T)baseValue.Multiply(baseValue);
            b /= 2;
        }
        return result;
    }
}

File: l3/z2/lib/DHSetup.java
package lib;

import java.util.ArrayList;
import java.util.List;
import java.util.Random;

public class DHSetup<T extends Field<T>> {
    private T generator;
    private final long characteristic;
    private final T instance;

    public DHSetup(Class<T> type) {
        try {
            this.instance = type.getDeclaredConstructor().newInstance();
            this.characteristic = instance.getCharacteristic();
        } catch (Exception e) {
            throw new RuntimeException("Error creating instance of " + type.getName(), e);
        }

        Random rd = new Random();
        List<Long> primeDivisors = calculatePrimeDivisors(characteristic - 1);

        boolean isValidGenerator = false;
        while (!isValidGenerator) {
            long candidateValue = 2 + rd.nextInt((int) (characteristic - 2));
            T candidate = instance.createInstance(candidateValue);

            isValidGenerator = true;
            for (long prime : primeDivisors) {
                T powerResult = power(candidate, (characteristic - 1) / prime);
                if (powerResult.getValue() == 1) {
                    isValidGenerator = false;
                    break;
                }
            }
            if (isValidGenerator) {
                generator = candidate;
            }
        }
    }

    public T getGenerator() {
        return generator;
    }

    public T power(T base, long exp) {
        T result = instance.createInstance(1);
        base = instance.createInstance(base.getValue());
        while (exp > 0) {
            if (exp % 2 == 1) {
                result = result.multiply(base);
            }
            base = base.multiply(base);
            exp /= 2;
        }
        return result;
    }

    private List<Long> calculatePrimeDivisors(long n) {
        List<Long> primeDivisors = new ArrayList<>();
        while (n % 2 == 0) {
            primeDivisors.add(2L);
            n /= 2;
        }
        for (long i = 3; i <= Math.sqrt(n); i += 2) {
            while (n % i == 0) {
                primeDivisors.add(i);
                n /= i;
            }
        }
        if (n > 2) {
            primeDivisors.add(n);
        }
        return primeDivisors;
    }
}
